smithy-aws-protocol-tests/model/rpcV2/empty-input-output.smithy

```diff
@@ -0,0 +1,173 @@
+        protocol: rpcv2,
+        documentation: "Body is empty and no Content-Type header if no response",
+        body: "",
+        bodyMediaType: "application/cbor",
+        headers: {
+            "smithy-protocol": "rpc-v2-cbor",
+        },
+        forbidHeaders: [
+            "Content-Type"
+        ]
+        code: 200,
+    },
+    {
+        id: "no_output_client_allows_accept",
+        protocol: rpcv2,
+        documentation: "Servers should allow the accept header to be set to the default content-type.",
+        headers: {
+            "smithy-protocol": "rpc-v2-cbor",
+            "Accept": "application/cbor",
+            "Content-Type": "application/cbor"
+        }
+        body: "",
+        code: 200,
+        appliesTo: "client",
+    },
+    {
+        id: "no_input_client_allows_empty_cbor",
+        protocol: rpcv2,
+        documentation: "Client should accept CBOR empty struct if no output",
+        headers: {
+            "smithy-protocol": "rpc-v2-cbor",
+            "Accept": "application/cbor",
+            "Content-Type": "application/cbor"
+        }
+        body: "v/8=",
+        code: 200,
+        appliesTo: "client",
+    }
+])
+operation NoInputOutput {}
+
+
+@httpRequestTests([
+    {
+        id: "empty_input",
+        protocol: rpcv2,
+        documentation: "When Input structure is empty we write CBOR equivalent of {}",
+        headers: {
+            "smithy-protocol": "rpc-v2-cbor",
+            "Accept": "application/cbor",
+            "Content-Type": "application/cbor"
+        },
+        forbidHeaders: [
+            "X-Amz-Target"
+        ]
+        method: "POST",
+        uri: "/service/aws.protocoltests.rpcv2.RpcV2Protocol/operation/EmptyInputOutput",
+        body: "v/8=",
+    },
+])
+@httpResponseTests([
+    {
+        id: "empty_output",
+        protocol: rpcv2,
+        documentation: "When output structure is empty we write CBOR equivalent of {}",
+        body: "",
+        bodyMediaType: "application/cbor",
+        headers: {
+            "smithy-protocol": "rpc-v2-cbor",
+            "Content-Type": "application/cbor"
+        }
+        code: 200,
+    },
+])
+operation EmptyInputOutput {
+    input: EmptyStructure,
+    output: EmptyStructure
+}
+
+@httpRequestTests([
+    {
+        id: "optional_input",
+        protocol: rpcv2,
+        documentation: "When input is empty we write CBOR equivalent of {}",
+        headers: {
+            "smithy-protocol": "rpc-v2-cbor",
+            "Accept": "application/cbor",
+            "Content-Type": "application/cbor"
+        },
+        forbidHeaders: [
+            "X-Amz-Target"
+        ]
+        method: "POST",
+        uri: "/service/aws.protocoltests.rpcv2.RpcV2Protocol/operation/OptionalInputOutput",
+        body: "v/8=",
+    },
+])
+@httpResponseTests([
+    {
+        id: "optional_output",
+        protocol: rpcv2,
+        documentation: "When output is empty we write CBOR equivalent of {}",
+        body: "",
```

Comment:
My implementation fails this test because it generates:

```rust
pub fn ser_optional_input_output_output(
    encoder: &mut ::aws_smithy_cbor::Encoder,
    input: &crate::output::OptionalInputOutputOutput,
) -> Result<(), ::aws_smithy_http::operation::error::SerializationError> {
    encoder.begin_map();
    if let Some(var_1) = &input.value {
        encoder.str("value").str(var_1.as_str());
    }
    encoder.end();
    Ok(())
}
```

So when no members are set, it writes an empty indefinite-length CBOR map (a body with two bytes: [0xbf, 0xff]) instead of an empty map.

I'd expect all implementations of the protocol to do what I'm doing here: it's costly to calculate the amount of set members in a structure in order to use a definite-length map or not write anything at all.

I'm also unsure about whether expecting nothing at all here is compliant, or desirable. A deserializer is going to expect a map; to accomodate for the case where nothing is on the wire it would have to (1) write the expectation fallibly, and (2) in the error case check that the decoder is at the end of the document, returning an empty structure.

---

The question about how to tell Smithy protocol tests that it should accept definite-length as well as indefinite-length encodings is still open and I don't know how to solve.

Reply:
> I don't see why can't both definite (0xa0) and indefinite-length map (0xbf,0xff) both represent an empty map.

That's valid and also not what your test is asserting. The CBOR equivalent of `{}` is a map type with zero elements, but we're asserting that no payload at all should be written in this case.